Add new checker len-as-condition

atodorov opened the pull request:
This new checker implements the following statement from PEP8:

```
For sequences, (strings, lists, tuples), use the fact that empty sequences are false.

Yes: if not seq:
     if seq:

No: if len(seq):
    if not len(seq):
```

This is still a work in progress mainly around lines 76-79 when we have a Compare statement. ATM I'm not handling statements like `0 < len(S) < 5` or `0 != len(S)`, etc. Advice is welcome on how to handle this and what sort of error conditions to consider valid.

Background: Recently I've been working with a code base which has lots of `if len(something) > 0` or `if len(something) != 0` and figured I'll write this patch.

atodorov commented:
NOTE: I'm also not sure how to fix the remaining pylint warnings. Need a hint.

rogalski commented:
@atodorov what's unclear with pylint errors? Your new checker simply found these violations in current pylint source. https://github.com/PyCQA/pylint/blob/master/pylint/config.py#L175 is one of errors, I'm sure rest follow the same pattern.

atodorov commented:
@rogalski thanks for the hint. I thought the pylint errors were somewhere in the code I wrote and didn't realize the new checker will run against the rest of the project. I've finalized the len_cherker, added some more tests and also updated the rest of the project where the new checker complained.

This PR is ready for review.

atodorov commented:
ping, any updates on this one ?

atodorov commented:
This is ready to be merged, please review.

PCManticore commented:
I didn't have time lately for doing a thorough review. In general, it looks good, but I will get back with a more thorough analysis in the following days.

rowillia commented on pylint/checkers/len.py:
```diff
@@ -0,0 +1,100 @@
+        # `not len(S)` must become `not S` regardless if the parent block
+        # is a test condition or something else (boolean expression)
+        # e.g. `if not len(S):`
+        if isinstance(node, UnaryOp) and node.op == 'not' and \
```
Here and below: Prefer implied line continuations - https://www.python.org/dev/peps/pep-0008/#multiline-if-statements

Also, it'd be great to indent the `_is_len_call` call.  Right now quickly glancing at the code it's easy to confuse whether or not that's part of the body. 

```python
        if (isinstance(node, UnaryOp) and node.op == 'not' and
                _is_len_call(node.operand)):
            self.add_message('len-as-condition', node=node)
```

rowillia commented on pylint/checkers/len.py:
```diff
@@ -0,0 +1,100 @@
+
+def _is_len_call(node):
+    """checks if @node is len(SOMETHING)"""
+    return isinstance(node, Call) and isinstance(node.func, Name) and node.func.name == 'len'
```
Can you also ensure `node` refers to the built-in `len`?

Unfortunately there's two methods that implement this check in the codebase ☹️ 

https://github.com/PyCQA/pylint/blob/2beb5e650d689d4903676b5ca6d82a1806bbbe79/pylint/checkers/refactoring.py#L342-L346
https://github.com/PyCQA/pylint/blob/23e59c80cfcba5878b7f634464a203017ab91187/pylint/checkers/python3.py#L44-L45

rowillia commented on pylint/test/functional/len_checks.py:
```diff
@@ -0,0 +1,30 @@
+    pass
+
+if 10 > len('TEST') > 1 > 0:
+    pass
```
Some other test case ideas:

```python
foo = len('TEST') or 42  # Should be fine

if 0 <= len('TEST') < 100:  # Should be fine
  pass

if z and len(x):  # Should fail ideally, but I realize that would mean having to walk the tree looking for `test` calls.
  pass

a = x and len(x)  # Should be fine
```

atodorov replied:
added

rowillia commented on pylint/checkers/len.py:
```diff
@@ -0,0 +1,100 @@
+        # a len(S) call is used inside a test condition
+        # could be if, while, assert or if expression statement
+        # e.g. `if len(S):`
+        if hasattr(node.parent, 'test') and _is_len_call(node):
```
Do you want this to be only for first order `len` calls or for complex ones as well?  e.g. should 

```
if z and len(x):
    pass
```

Get a warning?  If not, we might want to explicitly call that out in docs.

atodorov replied:
Good catch. I've made a slight modification to detect whether or not the `len()` call is nested inside BoolOp, which at some point may have a parent with the test attribute.

rowillia commented on pylint/checkers/len.py:
```diff
@@ -0,0 +1,100 @@
+        ops.extend(node.ops)
+        ops = list(itertools.chain(*ops))
+
+        for ops_idx in range(len(ops)-2):
```
Here and below: consider adding some space between the operators operands

rowillia commented on pylint/checkers/len.py:
```diff
@@ -0,0 +1,100 @@
+"""check for incorrect usage of len() in conditions
+"""
+import itertools
+from astroid.node_classes import Call, Const, Name, UnaryOp
```
Elsewhere in this codebase we generally use `node_classes` directly from `astroid` (it's `__init__.py` makes them accessable).   For example:

https://github.com/PyCQA/pylint/blob/c72d1070bc1dab7d0bcb45b5fe52b8f6dd4bba5e/pylint/checkers/classes.py#L223
or
https://github.com/PyCQA/pylint/blob/master/pylint/checkers/python3.py#L68

rowillia commented on pylint/checkers/len.py:
```diff
@@ -0,0 +1,100 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/master/COPYING
+
+"""check for incorrect usage of len() in conditions
```
Here and below: Capitalize first letter and add period at the end.

rowillia commented on pylint/checkers/len.py:
```diff
@@ -0,0 +1,100 @@
+            op_3 = ops[ops_idx+2]
+
+            # 0 ?? len()
+            if isinstance(op_1, Const) and (op_1.value == 0) and \
```
Consider pulling this out into a function like `_is_constant_zero` or something to reduce code duplication.

rowillia commented on pylint/checkers/len.py:
```diff
@@ -0,0 +1,100 @@
+    def visit_compare(self, node):
+        # compare nodes are trickier because the len(S) expression
+        # may be somewhere in the middle of the node
+        ops = [('', node.left)]
```
I don't quite follow this logic.  I believe that it's trying to consider `0 < len(x) < 10` by always setting this up as a 3 way comparison, right?  Seems to work but I'd love a quick comment as to how it works 😄

atodorov replied:
Added a comment in the code but basically this is because of the way astroid creates the Compare nodes. The left most operand is in node.left.

rowillia commented on doc/whatsnew/2.0.rst:
```diff
@@ -89,6 +89,29 @@ New checkers
+
+      if len(S) == 0:
+        pass
+
```
Consider adding an example of how to correctly handle this and linking to https://www.python.org/dev/peps/pep-0008/#programming-recommendations for our less experienced python developer friends 😄

atodorov commented:
@rowillia thanks for the review. I'm working my way through the comments and will reply where necessary.

PCManticore commented on doc/whatsnew/2.0.rst:
```diff
@@ -89,6 +89,51 @@ New checkers
+  we detect that a condition uses ``len(SEQUENCE)`` incorrectly. Instead
+  one could use ``if SEQUENCE`` or ``if not SEQUENCE``.
+
+  For instance, all of the examples below are incorrect:
```
I would not use incorrect here, rather something along the lines "all the examples below could be written in a more natural way".

PCManticore commented on pylint/checkers/len.py:
```diff
@@ -0,0 +1,113 @@
+
+"""Check for incorrect usage of len() in conditions."""
+import itertools
+import astroid
```
Separate stdlib from third party imports.

PCManticore replied:
Not sure I like this being a separate file. We tend to have checkers grouped by a general concept, such as refactoring or checkers related to classes. I would say this belongs into refactoring.py.

PCManticore commented on pylint/checkers/len.py:
```diff
@@ -0,0 +1,113 @@
+
+
+def _is_len_call(node):
+    """Checks if @node is len(SOMETHING)."""
```
Use ``node`` rather than ``@node``

PCManticore commented on pylint/checkers/len.py:
```diff
@@ -0,0 +1,113 @@
+            node.func.name == 'len')
+
+def _is_constant_zero(node):
+    return isinstance(node, astroid.Const) and (node.value == 0)
```
You can drop the parens for ``node.value == 0``.

PCManticore commented on pylint/checkers/len.py:
```diff
@@ -0,0 +1,113 @@
+    # messages
+    msgs = MSGS
+    priority = -2
+    # configuration options
```
These comments are superfluous.

PCManticore commented on pylint/checkers/len.py:
```diff
@@ -0,0 +1,113 @@
+
+            # we're finally out of any nested boolean operations so check if
+            # the parent is indeed an if, while, assert of if expression statement
+            if hasattr(parent, 'test'):
```
I prefer an explicit instance check, something as in ``isinstance(parent, (astroid.If, ...))```

PCManticore commented:
Looks pretty nice. Left a couple of minor comments.

atodorov commented:
@PCManticore - thanks for the review, updated and merged with `refactoring.py`

atodorov commented on pylint/checkers/base.py:
```diff
@@ -203,7 +203,7 @@ def _has_abstract_methods(node):
     with `abc` decorators.
     """
-    return len(utils.unimplemented_abstract_methods(node)) > 0
+    return bool(utils.unimplemented_abstract_methods(node))
```
I've made a small modification to the existing code. It will traverse the parent nodes until it reaches a node like `if` or `while` where the trithiness of the value is checked. It should be fine now.

rowillia commented on doc/whatsnew/2.0.rst:
```diff
@@ -89,6 +89,51 @@ New checkers
+      if not S:
+        pass
+
+      if S:
```
Nit: I would just leave the two examples, no need to repeat them 😄

rowillia commented on pylint/checkers/refactoring.py:
```diff
@@ -476,8 +478,118 @@ def visit_unaryop(self, node):
+
+    # configuration section name
+    name = 'len'
+    msgs = {'W1801': ('Do not use `len(SEQUENCE)` as condition value',
```
This should start with `C` as it's a coding convention (See https://pylint.readthedocs.io/en/latest/user_guide/output.html#source-code-analysis-section)

rowillia commented on pylint/checkers/refactoring.py:
```diff
@@ -476,8 +478,118 @@ def visit_unaryop(self, node):
+    return isinstance(node, astroid.Const) and node.value == 0
+
+def _node_is_test_condition(node):
+    # the node is an if, while, assert of if expression statement
```
If this is a docstring it should be formatted as such instead of as a comment.

rowillia commented on pylint/checkers/refactoring.py:
```diff
@@ -476,8 +478,118 @@ def visit_unaryop(self, node):
+
+    @utils.check_messages('len-as-condition')
+    def visit_unaryop(self, node):
+        # `not len(S)` must become `not S` regardless if the parent block
```
Convert to a Docstring

rowillia commented on pylint/checkers/refactoring.py:
```diff
@@ -476,8 +478,118 @@ def visit_unaryop(self, node):
+
+
+class LenChecker(checkers.BaseChecker):
+    """Checks for incorrect usage of len() inside conditions. pep8 states:
```
Super small nit:  Move `pep8 states:`... to next line to clean up one-line docstring for this method.  Also, `pep8` should be capitalized.

atodorov commented:
@rowillia updated & rebased

rowillia commented:
@atodorov Great, Thanks!  Appveyor failure looks unrelated.
